pkg/cpe/wfn.go

```diff
@@ -0,0 +1,209 @@
+package cpe
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+// Attribute is a type for enumerating the valid CPE attributes.
+type Attribute int
+
+//go:generate stringer -type Attribute -linecomment
+
+// These are the valid Attributes
+const (
+	Part      Attribute = iota // part
+	Vendor                     // vendor
+	Product                    // product
+	Version                    // version
+	Update                     // update
+	Edition                    // edition
+	SwEdition                  // sw_edition
+	TargetSW                   // target_sw
+	TargetHW                   // target_hw
+	Language                   // language
+	Other                      // other
+)
+
+// NumAttr is the number of attributes in a 2.3 WFN.
+const NumAttr = 11
+
+// NonASCII reports true if the rune is not ASCII.
+func nonASCII(r rune) bool {
+	return r >= unicode.MaxASCII
+}
+
+// Reserved reports true if the rune is in the "reserved" set for CPE
+// strings and needs quoting.
+func reserved(r rune) bool {
+	return !((r >= 0x30 && r <= 0x39) ||
+		(r >= 0x41 && r <= 0x5a) ||
+		(r >= 0x61 && r <= 0x7a) ||
+		r == '_')
+}
+
+// Validate is adapted from
+// https://nvlpubs.nist.gov/nistpubs/Legacy/IR/nistir7695.pdf section 5.3.2
+func validate(s string) error {
+	if !utf8.ValidString(s) {
+		return fmt.Errorf("cpe: string not valid utf8")
+	}
+	if strings.IndexFunc(s, nonASCII) != -1 {
+		return fmt.Errorf("cpe: string contains non-ASCII characters")
+	}
+	if strings.IndexFunc(s, unicode.IsSpace) != -1 {
+		return fmt.Errorf("cpe: string contains space characters")
+	}
+	// Special character rules:
+	// A single * means ANY and a quoted hyphen would get unquoted, which means
+	// NA, so those are disallowed as string values.
+	if s == "*" {
+		return fmt.Errorf("cpe: single asterisk MUST NOT be used by itself")
+	}
+	if s == "\\-" {
+		return fmt.Errorf("cpe: quoted hyphen MUST NOT be used by itself")
+	}
+	var (
+		// In 'escaped' state
+		esc = false
+		// position of last rune
+		last = len(s) - 1
+		// if in a run of ? and if we've seen a non-? character
+		qRun, atStart = false, true
+	)
+	for i, r := range s {
+		switch r {
+		case '*':
+			if esc { // if escaped, it's not special
+				break
+			}
+			if i != 0 && i != last {
+				return fmt.Errorf("cpe: invalid position for special character: %q at %d", r, i)
+			}
+		case '?':
+			if esc { // if escaped, it's not special
+				break
+			}
+			qRun = true
+		case '\\':
```

Comment:
why is this a double escape?

Reply:
need to match a single `\`; can't use backticks because that's a string literal, and single quote allows for escaping single quotes, so the backslash needs to be doubled to get a literal backslash.